internal/core/src/index/StringIndexMarisa.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include "index/StringIndexMarisa.h"
+#include "index/Utils.h"
+#include "index/Index.h"
+
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_io.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <knowhere/common/Utils.h>
+#include <pb/schema.pb.h>
+
+namespace milvus::scalar {
+
+#ifdef __linux__
+
+int64_t
+StringIndexMarisa::Size() {
+    return trie_.size();
+}
+
+void
+StringIndexMarisa::Build(size_t n, const std::string* values) {
+    if (built_) {
+        throw std::runtime_error("index has been built");
+    }
+
+    marisa::Keyset keyset;
+    {
+        // fill key set.
+        for (size_t i = 0; i < n; i++) {
+            keyset.push_back(values[i].c_str());
+        }
+    }
+
+    trie_.build(keyset);
+    fill_str_ids(n, values);
+    fill_offsets();
+
+    built_ = true;
+}
+
+BinarySet
+StringIndexMarisa::Serialize(const Config& config) {
+    auto uuid = boost::uuids::random_generator()();
+    auto uuid_string = boost::uuids::to_string(uuid);
+    auto file = std::string("/tmp/") + uuid_string;
+
+    auto fd = open(file.c_str(), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR | S_IXUSR);
+    trie_.write(fd);
```

Comment:
I think you should use "trie_.map()" or "trie_.mmap()" instead

Reply:
thx, I'll try this way.